### PR TITLE
prod-host-setup: recommendation on async-unsafe signal handlers

### PR DESCRIPTION
**Pull request by alindima**
# Reason for This PR

https://github.com/firecracker-microvm/firecracker/issues/2452

## Description of Changes

- add a note and recommendation in the prod-host-setup.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: docs/prod-host-setup.md [hunk 1]
```diff
@@ -59,6 +59,20 @@ logging, which degraded the snapshot restore time from 3ms to 8.5ms on
...
 generated host kernel logs, which were very slow to write.
 
+### Logging and signal handlers
+
+Firecracker installs custom signal handlers for some of the POSIX signals, such
+as SIGSEGV, SIGSYS, etc.
+The custom signal handlers used by Firecracker are not async-signal-safe, since
+they write logs and flush the metrics, which use locks for synchronization.
+There is a very low probability, that, when running the signal handler on a
+thread, that thread was already holding a lock to the log or metrics buffers.
```
Comment (georgepisaltu) [inline docs/prod-host-setup.md:69]: This sentence is hard to read, I'll suggest this but feel free to rephrase however you see fit.

```diff
- There is a very low probability, that, when running the signal handler on a
- thread, that thread was already holding a lock to the log or metrics buffers.
+ It is possible but very unlikely that the handler will intercept a signal on a 
+ thread which is already holding a lock to the log or metrics buffer.
```

File: docs/prod-host-setup.md [hunk 2]
```diff
@@ -59,6 +59,20 @@ logging, which degraded the snapshot restore time from 3ms to 8.5ms on
 `aarch64`. In this case, creating the tap device for snapshot restore
 generated host kernel logs, which were very slow to write.
 
+### Logging and signal handlers
+
+Firecracker installs custom signal handlers for some of the POSIX signals, such
+as SIGSEGV, SIGSYS, etc.
+The custom signal handlers used by Firecracker are not async-signal-safe, since
```
Comment (georgepisaltu) [inline docs/prod-host-setup.md:66]: If you meant to add a newline here, you need to add one more, like

```diff
- as SIGSEGV, SIGSYS, etc.
+ as SIGSEGV, SIGSYS, etc.
+ 
```

File: docs/prod-host-setup.md [hunk 3]
```diff
@@ -59,6 +59,20 @@ logging, which degraded the snapshot restore time from 3ms to 8.5ms on
...
+they write logs and flush the metrics, which use locks for synchronization.
+There is a very low probability, that, when running the signal handler on a
+thread, that thread was already holding a lock to the log or metrics buffers.
+This can result in a deadlock, where the specific Firecracker thread becomes
+unresponsive.
+While there is no security impact cause by the unlikely deadlock, we recommend
```
Comment (georgepisaltu) [inline docs/prod-host-setup.md:72]: Same as above regarding the newline.

Comment (georgepisaltu) [inline docs/prod-host-setup.md:72]:
```diff
- While there is no security impact cause by the unlikely deadlock, we recommend
+ While there is no security impact caused by the deadlock, we recommend
```

File: docs/prod-host-setup.md [hunk 4]
```diff
@@ -59,6 +59,20 @@ logging, which degraded the snapshot restore time from 3ms to 8.5ms on
...
+that customers have an overwatcher process on the host, that periodically looks
+for Firecracker processes that are un-responsive, and kills them, by SIGKILL.
```
Comment (georgepisaltu) [inline docs/prod-host-setup.md:74]: Do we have a recommendation for the specific method used to determine if the process is unresponsive? If we do, maybe it's worth mentioning here.

Also

```diff
- for Firecracker processes that are un-responsive, and kills them, by SIGKILL.
+ for Firecracker processes that are unresponsive and kills them with SIGKILL.
```

Comment (alindima) [reply docs/prod-host-setup.md]: There are two cases to think about:

**1. a thread is deadlocked on flushing metrics**

If another thread than the VMM is deadlocked on metrics, the VMM thread will also get deadlocked in at most 1 minute (which is the metric flushing period).
We are now sure that the VMM thread is deadlocked as well, if one of the threads was deadlocked.
To detect that the VMM is deadlocked, one can either have an agent inside the VM that sends some data periodically to an overwatcher, or simply have an overwatcher poll the instance state periodically (which will trigger some code on the VMM thread, which is deadlocked, so it will timeout).

**2. a thread is deadlocked on writing a log**

if the VMM thread is blocked logging, any of the approaches in the previous point would work (polling instance state or having an agent that uses an emulated device, like net, vsock).
If the API thread is blocked while logging, the approach with polling the instance state also works.
Let's think what happens is a vcpu is deadlocked. It must have performed a VMEXIT, because the vcpu was interrupted by a signal while logging. If a vcpu is blocked in a VMEXIT, sooner or later, a lock-up will be detected by the guest (if it has more than one vcpus). If it's the only vcpu of the VM, having an agent inside the VM will help a customer detect this.

Hopefully my thinking makes sense.
To summarise, I would recommend that the overwatcher does this (assuming that Firecracker metrics are enabled and the guest has a lockup detector):
1. Poll instance state
2. if step 1 succeeded, proceed to ping the guest over network/vsock.

These steps are to be repeated periodically, on a sensible period.

How does this sound?

Comment (alindima) [reply docs/prod-host-setup.md]: Note that the above makes sense to implement only if customers don't already have systems that already do this.

Comment (alindima) [reply docs/prod-host-setup.md]: @firecracker-microvm/compute-capsule what are you thoughts?

Comment (georgepisaltu) [reply docs/prod-host-setup.md]: > Note that the above makes sense to implement only if customers don't already have systems that already do this.

This is a very good point and I think we should start our recommendation with this if we decide to include the suggestion.

> if step 1 succeeded, proceed to ping the guest over network/vsock

I don't think forcing customers to use the net/vsock devices to make sure the VM is responsive is good, as they are both optional devices. I thought we could go a simpler route and just drop a quick suggestion, like for example trying to run any API request which interacts with the VMM emulation thread (a `VmmAction`, like a pause request). If there is an interval of time of up to 1 minute in which these requests might work even though the VM is doomed, I don't see how we present this as a reasonable option. Maybe just monitor the VMM thread with a tool for idle time? I have no clear solution, so I am fine with not including this.

**Review by acatangiu — APPROVED** [review]

**Review by georgepisaltu — APPROVED** [review]
